packages/store/src/lib/AtomMap.ts

```diff
@@ -0,0 +1,201 @@
+		}
+		return valueAtom.get() !== UNINITIALIZED
+	}
+
+	__unsafe__hasWithoutCapture(key: K): boolean {
+		const valueAtom = this.valueMap.get(key)
+		if (!valueAtom) return false
+		return valueAtom.__unsafe__getWithoutCapture() !== UNINITIALIZED
+	}
+
+	set(key: K, value: V) {
+		transact(() => {
+			const existingAtom = this.valueMap.get(key)
+			if (existingAtom) {
+				existingAtom.set(value)
+			} else {
+				this.valueMap.set(key, atom(`${this.name}:${String(key)}`, value))
+				this.presentKeysAtom.update((keys) => {
+					const newKeys = new Set(keys)
+					newKeys.add(key)
+					return newKeys
+				})
+			}
+		})
+		return this
+	}
+
+	update(key: K, updater: (value: V) => V) {
+		return transact(() => {
+			const valueAtom = this.valueMap.get(key)
+			if (!valueAtom) {
+				throw new Error(`AtomMap: key ${key} not found`)
+			}
+			const value = valueAtom.__unsafe__getWithoutCapture()
+			if (value === UNINITIALIZED) {
+				throw new Error(`AtomMap: key ${key} not found`)
+			}
+			valueAtom.set(updater(value))
+		})
+	}
+
+	delete(key: K) {
+		return transact(() => {
+			const valueAtom = this.valueMap.get(key)
+			if (!valueAtom) {
+				return false
+			}
+			valueAtom.set(UNINITIALIZED)
+			this.presentKeysAtom.update((keys) => {
+				const newKeys = new Set(keys)
+				newKeys.delete(key)
+				return newKeys
+			})
+			this.valueMap.delete(key)
+			return true
+		})
+	}
+
+	deleteMany(keys: Iterable<K>): [K, V][] {
+		return transact(() => {
+			let newPresentKeys
+			const deleted: [K, V][] = []
+
+			for (const key of keys) {
+				const valueAtom = this.valueMap.get(key)
+				if (!valueAtom) continue
+				const oldValue = valueAtom.get()
+				if (oldValue === UNINITIALIZED) continue
+
+				deleted.push([key, oldValue])
+
+				valueAtom.set(UNINITIALIZED)
+				if (!newPresentKeys) {
+					newPresentKeys = new Set(this.presentKeysAtom.__unsafe__getWithoutCapture())
+				}
+				newPresentKeys.delete(key)
+				this.valueMap.delete(key)
+			}
+
+			if (newPresentKeys) {
+				this.presentKeysAtom.set(newPresentKeys)
+			}
+
+			return deleted
+		})
+	}
+
+	clear() {
+		return transact(() => {
+			for (const valueAtom of this.valueMap.values()) {
+				valueAtom.set(UNINITIALIZED)
+			}
+			this.presentKeysAtom.set(new Set())
+			this.valueMap.clear()
+		})
+	}
+
+	*entries(): Generator<[K, V], undefined, unknown> {
+		for (const key of this.presentKeysAtom.get()) {
+			const value = this.getAtom(key)!.get()
+			if (value === UNINITIALIZED) {
+				continue
+			}
+			yield [key, value]
+		}
+	}
+
+	*keys(): Generator<K, undefined, unknown> {
+		for (const key of this.presentKeysAtom.get()) {
+			const value = this.getAtom(key)!.get()
+			if (value === UNINITIALIZED) {
+				continue
+			}
+			yield key
+		}
+	}
+
+	*values(): Generator<V, undefined, unknown> {
+		for (const key of this.presentKeysAtom.get()) {
+			const value = this.getAtom(key)!.get()
+			if (value === UNINITIALIZED) {
+				continue
+			}
+			yield value
+		}
+	}
```

Comment:
unlike normal js maps these iterators won't be tolerant of updates (particularly insertions) because they iterate over the presentKeysAtom rather than the underlying JS Map.

Maybe they should deference the presentKeysAtom to get reactivity, and then iterate over this.valueMap.values() etc to make sure the iterator gets values that were inserted during iteration like a normal js map?